File: clang/ads/stack.c
#include <stdio.h>
#include <stdbool.h>
#include <stdbool.h>

#define MAX_SIZE 100

typedef struct {
  int top;
  int array[MAX_SIZE];
} Stack;

int   pop(Stack* stack);
int   peek(Stack* stack);
bool  is_full(Stack* stack);
bool  is_empty(Stack* stack);
void  init(Stack* stack);
void  push(Stack* stack, int value);

void init(Stack* stack){
  stack->top = -1;
}

bool is_empty(Stack* stack){
  return stack->top == -1;
}

bool is_full(Stack* stack){
  return stack->top == MAX_SIZE - 1;
}

void push(Stack* stack, int value){
  if (is_full(stack)){
    printf("Stack Overflow\n");
    return;
  }

  stack->array[++stack->top] = value;
  printf("Pushed %d onto the stack\n", value);
}

int pop(Stack* stack){
  if (is_empty(stack)){
    printf("Stack Overflow\n");
    return -1;
  }

  int popped = stack->array[stack->top];
  stack->top--;
  printf("Popped %d from the stack\n", popped);

  return popped;
}

int peek(Stack* stack){
  if (is_empty(stack)){
    printf("Stack is empty\n");
    return -1;
  }
  return stack->array[stack->top];
}

File: nim/ads/stack.nim
const ARRAY_SIZE = 10

type Stack = object
  top: int
  arr: array[ARRAY_SIZE, int]

proc newStack(): Stack =
  Stack(top: -1)

proc isEmpty(stack: Stack): bool =
  stack.top == -1

proc isFull(stack: Stack): bool =
  stack.top == (ARRAY_SIZE - 1)

proc push(stack: var Stack, value: int) =
  if (stack.isFull()):
    echo "Stack Overflow"
    return

  let topp = stack.top + 1
  stack.top += 1
  stack.arr[topp] = value
  echo "Pushed " & $value & " onto the stack"

proc pop(stack: var Stack) =
  if (stack.isEmpty()):
    echo "Stack Overflow"
    return

  let popped = stack.arr[stack.top]
  echo "Popped " & $popped & " from the stack"
  stack.top -= 1

proc peek(stack: var Stack): int =
  if (stack.isEmpty()):
    echo "Stack is empty"
    return

  let popped = stack.arr[stack.top]
  echo "Peeked " & $popped & " from the stack"
  stack.arr[stack.top]

proc print(stack: Stack) =
  for i in countdown(stack.top, 0):
    echo stack.arr[i]

proc main() =
  var stack = newStack()
  stack.push(10)
  stack.push(20)
  stack.push(30)
  stack.push(40)
  stack.push(50)
  stack.push(60)
  stack.push(70)
  stack.push(80)
  stack.push(90)

  echo()
  stack.pop()
  stack.pop()
  stack.pop()
 
  echo()
  discard stack.peek()
  discard stack.peek()
  discard stack.peek()

  echo()
  stack.print()
  echo()

when isMainModule:
  main()
